core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -526,22 +526,20 @@ public boolean apply(@Nullable RecoveryStatus status) {
                         // closing the shard will also cancel any ongoing recovery.
                         indexService.removeShard(shardRouting.id(), "removing shard (recovery source node changed)");
                         shardHasBeenRemoved = true;
-
                     }
                 }
-                if (shardHasBeenRemoved == false && (shardRouting.equals(indexShard.routingEntry()) == false || shardRouting.version() > indexShard.routingEntry().version())) {
-                    if (shardRouting.primary() && indexShard.routingEntry().primary() == false && shardRouting.initializing() && indexShard.allowsPrimaryPromotion() == false) {
-                        logger.debug("{} reinitialize shard on primary promotion", indexShard.shardId());
-                        indexService.removeShard(shardId, "promoted to primary");
-                    } else {
-                        // if we happen to remove the shardRouting by id above we don't need to jump in here!
-                        indexShard.updateRoutingEntry(shardRouting, event.state().blocks().disableStatePersistence() == false);
-                    }
+
+                if (shardHasBeenRemoved == false) {
+                    // shadow replicas do not support primary promotion. The master would reinitialize the shard, giving it a new allocation, meaning we should be there.
+                    // nocommit: double check check this
```

Comment:
Okay, I tested this locally, it looks like it works differently with the line:

```
[2015-07-22 08:54:10,744][DEBUG][org.elasticsearch.indices.cluster] [node_t1] [test][0] removing shard (different instance of it allocated on this node, current [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [R], v[4], s[STARTED], a[id=93HrpsfMRWybeCn-NxhBjw]], global [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [P], v[6], s[INITIALIZING], a[id=g-oZ-7uiRTmDxgLnTdQHjg], unassigned_info[[reason=REINITIALIZED], at[2015-07-22T14:54:10.743Z]]])[2015-07-22 08:54:10,744][DEBUG][org.elasticsearch.indices.cluster] [node_t1] [test][0] removing shard (different instance of it allocated on this node, current [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [R], v[4], s[STARTED], a[id=93HrpsfMRWybeCn-NxhBjw]], global [[test][0], node[nNCDLISFRu2ncr-QCiYPgQ], [P], v[6], s[INITIALIZING], a[id=g-oZ-7uiRTmDxgLnTdQHjg], unassigned_info[[reason=REINITIALIZED], at[2015-07-22T14:54:10.743Z]]])
```

Is my understanding correct?